presto-main/src/main/java/com/facebook/presto/metadata/ParametricScalar.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.operator.scalar.ScalarFunctionImplementation;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.metadata.FunctionType.SCALAR;
+import static java.util.Objects.requireNonNull;
+
+public abstract class ParametricScalar
+        implements ParametricFunction
+{
+    private final Signature signature;
+
+    protected ParametricScalar(String name, List<TypeParameter> typeParameters, String returnType, List<String> argumentTypes)
+    {
+        this(name, typeParameters, returnType, argumentTypes, false);
+    }
+
+    protected ParametricScalar(String name, List<TypeParameter> typeParameters, String returnType, List<String> argumentTypes, boolean variableArity)
+    {
+        requireNonNull(name, "name is null");
+        requireNonNull(typeParameters, "typeParameters is null");
+        requireNonNull(returnType, "returnType is null");
+        requireNonNull(argumentTypes, "argumentTypes is null");
+        this.signature = new Signature(name, SCALAR, ImmutableList.copyOf(typeParameters), returnType, ImmutableList.copyOf(argumentTypes), variableArity);
+    }
+
+    @Override
+    public final Signature getSignature()
+    {
+        return signature;
+    }
+
+    public abstract ScalarFunctionImplementation specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry);
```

Comment:
Ya, we can definitely consider that. I don't want to put it in this PR, as this is just cleaning up some legacy code. Why doesn't the current interface work for you though? In your ParametricFunction implementation you should be able to expand the `types` and `arity` to a list, if you need to, by inserting them into the `List` in order and replicating the last parameter until `arity` is fulfilled. I'm working on a new framework for writing parametric functions using annotations that does something very similar